devito/types/basic.py

```diff
@@ -724,9 +733,22 @@ def dimensions(self):
         """Tuple of Dimensions representing the object indices."""
         return self._dimensions
 
+    @property
+    def _eval_deriv(self):
+        return self
+
+    @property
+    def on_grid(self):
```

Comment:
why do you need to check the extra attribute `on_grid` ? can you not just check the staggered mask?

Reply:
Checking the staggered mask requires to remove offsets from FD, then remove staggered offset then check it's on the grid as below which is way more expensive.

Reply:
can you not save it as a flag at construction time, ie at some point in `__new__` or `__init__` or `__init_finalize__` , where you have more raw information?

I guess I should just spend some serious time to actually understand what's really going on, rather than thinking on the level of individual commits...